pydantic/typing.py

```diff
@@ -128,6 +128,54 @@ def get_args(tp: Type[Any]) -> Tuple[Any, ...]:
         return typing_get_args(tp) or getattr(tp, '__args__', ()) or generic_get_args(tp)
 
 
+if sys.version_info >= (3, 9):
+    from typing import Annotated
+else:
+    if TYPE_CHECKING:
+        from typing_extensions import Annotated, _AnnotatedAlias
+    else:  # due to different mypy warnings raised during CI for python 3.7 and 3.8
+        try:
+            from typing_extensions import Annotated
+
+            try:
+                from typing_extensions import _AnnotatedAlias
+            except ImportError:
+                # py3.6 typing_extensions doesn't have _AnnotatedAlias, but has AnnotatedMeta, which
+                # will satisfy our `isinstance` checks.
+                from typing_extensions import AnnotatedMeta as _AnnotatedAlias
+        except ImportError:
+            # Create mock Annotated/_AnnotatedAlias values distinct from `None`, which is a valid
+            # `get_origin` return value.
+            class _FalseMeta(type):
+                # Allow short circuiting with "Annotated[...] if Annotated else None".
+                def __bool__(cls):
+                    return False
+
+                # Give a nice suggestion for unguarded use
+                def __getitem__(cls, key):
+                    raise RuntimeError('Annotated is not supported, please `pip install typing-extensions`.')
+
+            class Annotated(metaclass=_FalseMeta):
+                pass
+
+            class _AnnotatedAlias(metaclass=_FalseMeta):
+                pass
+
+    # Our custom get_{args,origin} for <3.8 and the builtin 3.8 get_{args,origin} don't recognize
+    # typing_extensions.Annotated, so wrap them to short-circuit. We still want to use our wrapped
+    # get_origins defined above for non-Annotated data.
+
+    def get_args(tp: Type[Any], _get_args=get_args) -> Type[Any]:
```

Comment:
1/3: We need the "Annotated aware" version in all places as the "default".

2: This probably gets tricky without larger "refactor now that typing-extensions is always available" changes, I think I'd have to add this logic to 3 versions of `get_args` and 2 versions of `get_origin` (the ones defined above) vs just wrapping whatever is decided above. Let me know if you'd like to just dup the check in those or refactor a bit.

Reply:
I was able to refactor the `get_args`/`get_origin` stuff a bit with `typing_extensions`, but it's not a panacea:
- those helper funcs aren't available on 3.6 so I kept the old "polyfills"
- I had to [remove](https://github.com/samuelcolvin/pydantic/pull/2147/files#diff-ada30799baaf47a93bb7fb7f732f34190e74564adc3e43d86924e8f354aadda8) two [tests](https://github.com/samuelcolvin/pydantic/pull/2147/files#diff-3a6553ad9b2043abd78833fc887feee49795abbc771b02266662d98676f4a113) using a bare `Callable` (the `typing_extensions` functions seem to expect the `Callable` generics to be `[...]`). Might get just as messy again to add the shim
- Looks like this makes the "test compiled without deps" fail (since we now import `typing_extensions`). Does this need to be tweaked somehow?

Reply:
Given that ^ messiness, I've pulled those changes (will reserve them for a future typing_extensions PR). How does [the latest commit](https://github.com/samuelcolvin/pydantic/pull/2147/commits/5bf718ab93e5c88c64a1c455e0f57fe75825d147) look?